client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java

```diff
@@ -0,0 +1,153 @@
+    this.scheduler = scheduler;
+    exceptionReporter = reporter;
+    this.maxEventsToFetch = maxEventsToFetch;
+  }
+
+  @Override
+  public void run() {
+    int failures = 0;
+    LOG.info(reduce + " Thread started: " + getName());
+
+    try {
+      while (!stopped && !Thread.currentThread().isInterrupted()) {
+        try {
+          int numNewMaps = getMapCompletionEvents();
+          failures = 0;
+          if (numNewMaps > 0) {
+            LOG.info(reduce + ": " + "Got " + numNewMaps + " new map-outputs");
+          }
+          LOG.debug("GetMapEventsThread about to sleep for " + SLEEP_TIME);
+          if (!Thread.currentThread().isInterrupted()) {
+            Thread.sleep(SLEEP_TIME);
+          }
+        } catch (InterruptedException e) {
+          LOG.info("EventFetcher is interrupted.. Returning");
+          return;
+        } catch (IOException ie) {
+          LOG.info("Exception in getting events", ie);
+          // check to see whether to abort
+          if (++failures >= MAX_RETRIES) {
+            throw new IOException("too many failures downloading events", ie);
+          }
+          // sleep for a bit
+          if (!Thread.currentThread().isInterrupted()) {
+            Thread.sleep(RETRY_PERIOD);
+          }
+        }
+      }
+    } catch (InterruptedException e) {
+      return;
+    } catch (Throwable t) {
+      exceptionReporter.reportException(t);
+      return;
+    }
+  }
+
+  public void shutDown() {
+    this.stopped = true;
+    interrupt();
+    try {
+      join(5000);
+    } catch(InterruptedException ie) {
+      LOG.warn("Got interrupted while joining " + getName(), ie);
+    }
+  }
+
+  /**
+   * Queries the {@link TaskTracker} for a set of map-completion events
+   * from a given event ID.
+   * @throws IOException
+   */
+  protected int getMapCompletionEvents()
+    throws IOException, InterruptedException {
+
+    int numNewMaps = 0;
+    TaskCompletionEvent events[] = null;
+
+    do {
+      MapTaskCompletionEventsUpdate update =
+        umbilical.getMapCompletionEvents(
+          (org.apache.hadoop.mapred.JobID)reduce.getJobID(),
+          fromEventIdx,
+          maxEventsToFetch,
+          (org.apache.hadoop.mapred.TaskAttemptID)reduce);
+      events = update.getMapTaskCompletionEvents();
+      LOG.debug("Got " + events.length + " map completion events from " +
+        fromEventIdx);
+
+      assert !update.shouldReset() : "Unexpected legacy state";
+
+      // Update the last seen event ID
+      fromEventIdx += events.length;
+
+      // Process the TaskCompletionEvents:
+      // 1. Save the SUCCEEDED maps in knownOutputs to fetch the outputs.
+      // 2. Save the OBSOLETE/FAILED/KILLED maps in obsoleteOutputs to stop
+      //    fetching from those maps.
+      // 3. Remove TIPFAILED maps from neededOutputs since we don't need their
+      //    outputs at all.
+      for (TaskCompletionEvent event : events) {
+        scheduler.resolve(event);
```

Comment:
We don't think we should add the concept `scheduler`.

Reply:
scheduler determines which task are required, and upload the progress.

Reply:
It's not neccessary concept. We can write a method to implement the function.

Reply:
scheduler has been removed.